bubuku/features/rebalance/change.py

```diff
@@ -0,0 +1,248 @@
+            # Get broker objects
+            source_broker, target_broker = brokers[top_candidate[0]], brokers[top_candidate[1]]
+
+            # Selecting best partition to move (It is better to just swap leadership, instead of copying data)
+            selected_partition = None
+            source_partitions = source_broker.list_partitions(topic, False)
+            target_partitions = target_broker.list_partitions(topic, True)
+            for partition in source_partitions:
+                if partition in target_partitions:
+                    selected_partition = partition
+                    break
+            # Failed to find optimized partition for swapping, taking any.
+            selected_partition = source_partitions[0] if selected_partition is None else selected_partition
+            topic_partition = (topic, selected_partition)
+            # Update brokers state
+            target_broker.accept_leader(source_broker, topic_partition)
+            # add transfer to queue
+            queue_list.append((topic_partition, source_broker.broker_id, target_broker.broker_id))
+            # Remove empty lists
+            for bp in [bp for bp, wl in candidates.items() if
+                       not brokers[bp[0]].have_extra_leaders() or not brokers[bp[1]].have_less_leaders()]:
+                del candidates[bp]
+        return queue_list
+
+    def _load_data(self):
+        self.broker_distribution = {id_: BrokerDescription(id_) for id_ in self.broker_ids}
+        self.source_distribution = {(topic, partition): replicas for topic, partition, replicas in
+                                    self.zk.load_partition_assignment()}
+        for topic_partition, replicas in self.source_distribution.items():
+            if not replicas:
+                continue
+            first = True
+            for replica in replicas:
+                if replica not in self.broker_distribution:
+                    self.broker_distribution[replica] = BrokerDescription(replica)
+                self.broker_distribution[replica].add_partition(first, topic_partition)
+                first = False
+        active_brokers = [broker for id_, broker in self.broker_distribution.items() if id_ in self.broker_ids]
+        distribute(
+            sum(b.get_leader_count() for b in self.broker_distribution.values()),
+            active_brokers, BrokerDescription.get_leader_count, BrokerDescription.set_leader_expectation)
+        # Here is one trick. Across all the code replica list does not contain leaders.
+        # But in order to balance correctly (max(diff(replica+leaders) <= 1) we should add leaders and substract it
+        # again
+        distribute(
+            sum((b.get_replica_count() + b.get_leader_count()) for b in self.broker_distribution.values()),
+            active_brokers,
+            BrokerDescription.get_replica_count,
+            lambda x, y: x.set_replica_expectation(y - x.get_expected_leaders()))
```

Comment:
what's `x` and `y`? If I understood the commentary correctly that it's just "undoing" something. With the goal of having the distribution of replicas correctly you need to subtract the amount of leaders. Maybe it would be more clear to remove this `lambda` as parameter and just iterate over the collection again but this time removing items, so you could have an intermediate collection with the "all partitions distribution" and you could obtain the "replicas distribution" by subtracting the "leader distribution". That would be easier to understand for me.